src/protocol/boolean/random_bits.rs

```diff
@@ -0,0 +1,170 @@
+    ) -> Result<(Vec<Replicated<F>>, Replicated<F>), BoxError> {
+        // We assume the bit length we operate in would not exceed 255
+        #[allow(clippy::cast_possible_truncation)]
+        let l = F::Integer::BITS as u8;
+
+        // Generate a pair of random numbers. We'll use these numbers as
+        // the source of `l`-bit long uniformly random sequence of bits.
+        let r_bits: (F, F) = ctx
+            .narrow(&Step::RandomFields)
+            .prss()
+            .generate_fields(record_id);
+
+        // Same here. For now, 64 bits is enough for our F_p
+        #[allow(clippy::cast_possible_truncation)]
+        let xor_shares = XorShares {
+            num_bits: l,
+            packed_bits_left: r_bits.0.as_u128() as u64,
+            packed_bits_right: r_bits.1.as_u128() as u64,
+        };
+
+        // Convert each bit to secret sharings of that bit in the target field
+        let c = ctx.narrow(&Step::ConvertShares);
+        let futures = (0..l).map(|i| {
+            // again, we don't expect our prime field to be > 2^64
+            #[allow(clippy::cast_possible_truncation)]
+            let c = c.narrow(&BitOpStep::Step(i as usize));
+            async move {
+                ConvertShares::new(xor_shares)
+                    .execute_one_bit(c, record_id, i as u8)
+                    .await
+            }
+        });
+        let r_b = try_join_all(futures).await?;
+
+        // Compute `[r_p]` by `Σ 2^i * [r_i]_B`
+        #[allow(clippy::cast_possible_truncation)]
+        let r_p: Replicated<F> = r_b
+            .iter()
+            .enumerate()
+            .fold(Replicated::ZERO, |acc, (i, x)| {
+                acc + &(x.clone() * F::from(2_u128.pow(i as u32)))
+            });
+
+        Ok((r_b, r_p))
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    RandomFields,
+    ConvertShares,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::RandomFields => "random_fields",
+            Self::ConvertShares => "convert_shares",
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::RandomBits;
+    use crate::{
+        error::BoxError,
+        ff::{Field, Fp31, Fp32BitPrime},
+        protocol::{context::ProtocolContext, QueryId, RecordId},
+        secret_sharing::Replicated,
+        test_fixture::{
+            bits_to_field, make_contexts, make_world, validate_and_reconstruct, TestWorld,
+        },
+    };
+    use futures::future::try_join_all;
+    use rand::{distributions::Standard, prelude::Distribution};
+
+    const TEST_TRIES: usize = 100;
+
+    async fn random_bits<F: Field>(
+        ctx: [ProtocolContext<'_, Replicated<F>, F>; 3],
+        record_id: RecordId,
+    ) -> Result<(Vec<F>, F), BoxError>
+    where
+        Standard: Distribution<F>,
+    {
+        let [c0, c1, c2] = ctx;
+
+        // Execute
+        let result = try_join_all(vec![
```

Comment:
```suggestion
        let result = try_join_all([
```

No need to allocate for this.

Reply:
That said, I've been considering whether we do a try_join3() wrapper for our code.  That might be worth doing.